File: python/src/opperai/clients/traces.py
"""Opper SDK — Traces Client."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from .._base_client import BaseClient
from ..types import (
    GetTraceResponse,
    ListTracesItem,
    ListTracesResponse,
    RequestOptions,
    TraceSpan,
)


class TracesClient:
    """Client for the Traces API endpoints."""

    def __init__(self, client: BaseClient) -> None:
        self._client = client

    def list(
        self,
        *,
        limit: int | None = None,
        offset: int | None = None,
        name: str | None = None,
        options: RequestOptions | None = None,
    ) -> ListTracesResponse:
        query: dict[str, Any] = {}
        if limit is not None:
            query["limit"] = limit
        if offset is not None:
            query["offset"] = offset
        if name is not None:
            query["name"] = name
        data = self._client._get("/v3/traces", query=query, options=options)
        items = [ListTracesItem(**item) for item in (data or {}).get("data", [])]
        meta = (data or {}).get("meta", {})
        return ListTracesResponse(data=items, meta=meta)

    async def list_async(
        self,
        *,
        limit: int | None = None,
        offset: int | None = None,
        name: str | None = None,
        options: RequestOptions | None = None,
    ) -> ListTracesResponse:
        query: dict[str, Any] = {}
        if limit is not None:
            query["limit"] = limit
        if offset is not None:
            query["offset"] = offset
        if name is not None:
            query["name"] = name
        data = await self._client._get_async("/v3/traces", query=query, options=options)
        items = [ListTracesItem(**item) for item in (data or {}).get("data", [])]
        meta = (data or {}).get("meta", {})
        return ListTracesResponse(data=items, meta=meta)

    def get(self, id: str, *, options: RequestOptions | None = None) -> GetTraceResponse:
        data = self._client._get(f"/v3/traces/{quote(id, safe='')}", options=options)
        spans = [TraceSpan(**s) for s in (data or {}).get("spans", [])]
        return GetTraceResponse(
            id=data.get("id", ""),
            name=data.get("name"),
            span_count=data.get("span_count", 0),
            spans=spans,
            start_time=data.get("start_time"),
            end_time=data.get("end_time"),
            duration_ms=data.get("duration_ms"),
            status=data.get("status"),
        )

    async def get_async(self, id: str, *, options: RequestOptions | None = None) -> GetTraceResponse:
        data = await self._client._get_async(f"/v3/traces/{quote(id, safe='')}", options=options)
        spans = [TraceSpan(**s) for s in (data or {}).get("spans", [])]
        return GetTraceResponse(
            id=data.get("id", ""),
            name=data.get("name"),
            span_count=data.get("span_count", 0),
            spans=spans,
            start_time=data.get("start_time"),
            end_time=data.get("end_time"),
            duration_ms=data.get("duration_ms"),
            status=data.get("status"),
        )

    def delete(self, id: str, *, options: RequestOptions | None = None) -> None:
        self._client._delete(f"/v3/traces/{quote(id, safe='')}", options=options)

    async def delete_async(self, id: str, *, options: RequestOptions | None = None) -> None:
        await self._client._delete_async(f"/v3/traces/{quote(id, safe='')}", options=options)

File: typescript/src/clients/traces.ts
import { BaseClient } from "../client-base.js";
import type {
  GetTraceResponse,
  ListTracesParams,
  ListTracesResponse,
  RequestOptions,
} from "../types.js";

/**
 * Client for the Traces API endpoints.
 * Provides methods to list, get, and delete traces.
 */
export class TracesClient extends BaseClient {
  /**
   * List traces with pagination.
   * GET /v3/traces
   */
  async listTraces(
    params?: ListTracesParams,
    options?: RequestOptions,
  ): Promise<ListTracesResponse> {
    return this._get<ListTracesResponse>(
      "/v3/traces",
      {
        limit: params?.limit,
        offset: params?.offset,
        name: params?.name,
      },
      options,
    );
  }

  /**
   * Get a trace with all its spans.
   * GET /v3/traces/{id}
   */
  async getTrace(id: string, options?: RequestOptions): Promise<GetTraceResponse> {
    const res = await this._get<{ data: GetTraceResponse }>(
      `/v3/traces/${encodeURIComponent(id)}`,
      undefined,
      options,
    );
    return res.data;
  }

  /**
   * Delete a trace and all its spans.
   * DELETE /v3/traces/{id}
   */
  async deleteTrace(id: string, options?: RequestOptions): Promise<void> {
    return this._delete<void>(`/v3/traces/${encodeURIComponent(id)}`, options);
  }
}
